src/main/java/net/neoforged/neoforge/fluids/crafting/FluidIngredient.java

```diff
@@ -0,0 +1,292 @@
+     * <p>
+     * For compatibility reasons, implementations should follow the same guidelines
+     * as for custom item ingredients, i.e.:
+     * <ul>
+     * <li>These stacks are generally used for display purposes, and need not be exhaustive or perfectly accurate.</li>
+     * <li>An exception is ingredients that {@linkplain #isSimple() are simple},
+     * for which it is important that the returned stacks correspond exactly to all the accepted {@link Fluid}s.</li>
+     * <li>At least one stack should always be returned, otherwise the ingredient may be considered {@linkplain #hasNoFluids() accidentally empty}.</li>
+     * <li>The ingredient should try to return at least one stack with each accepted {@link Fluid}.
+     * This allows mods that inspect the ingredient to figure out which stacks it might accept.</li>
+     * </ul>
+     *
+     * @return a stream of all fluid stacks this ingredient accepts.
+     *         <p>
+     *         Note: No guarantees are made as to the amount of the fluid,
+     *         as FluidIngredients are generally not meant to match by amount
+     *         and these stacks are mostly used for display.
+     *         <p>
+     * @see ICustomIngredient#getItems()
+     */
+    protected abstract Stream<FluidStack> generateStacks();
+
+    public abstract boolean isSimple();
+
+    /**
+     * {@return The type of this fluid ingredient.}
+     *
+     * <p>The type <b>must</b> be registered to {@link NeoForgeRegistries#FLUID_INGREDIENT_TYPES}.
+     */
+    public abstract FluidIngredientType<?> getType();
+
+    /**
+     * Checks if this ingredient is <b>explicitly empty</b>, i.e.
+     * equal to {@link EmptyFluidIngredient#INSTANCE}.
+     * <p> Note: This does <i>not</i> return true for "accidentally empty" ingredients,
+     * including compound ingredients that are explicitly constructed with no children
+     * or intersection / difference ingredients that resolve to an empty set.
+     *
+     * @return {@code true} if this ingredient is {@link #empty()}, {@code false} otherwise
+     */
+    public boolean isEmpty() {
+        return this == empty();
+    }
+
+    /**
+     * Checks if this ingredient matches no fluids, i.e. either:
+     * <ul>
+     * <li>is equal to {@link #empty()},</li>
+     * <li>resolves to an empty list of fluids</li>
+     * </ul>
+     * <p>
+     * Note that this method explicitly has the potential to <b>resolve</b>
+     * the ingredient; if this is not desired, you will need to check for
+     * emptiness another way!
+     *
+     * @return {@code true} if this ingredient matches no fluids, {@code false} otherwise
+     * @implNote If it is possible for your ingredient to return a "hint" on
+     *           whether it is empty without resolving, you should override this method
+     *           with a custom implementation.
+     * @see #isEmpty()
+     */
+    public boolean hasNoFluids() {
```

Comment:
This should be final as well, and it should be sufficient to check for `getStacks().length == 0`.

Reply:
Why should this be final actually? I would that if we can avoid early dissolution, i.e. give a "hint" on emptiness like the Javadoc mentions, we should do that instead of calling getStacks

Reply:
You can't avoid dissolution in this method so why bother trying for a limited subset of the ingredients? The caller has no way to know if it is safe or not to dissolve in this method, therefore it should always be assumed unsafe to call too early.

Reply:
Besides, leaving helper methods like this to be implemented makes it possible to introduce mistakes, and since it relies on dissolution the base implementation should be fast enough already.